### PR TITLE
Shard editing command

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

Allow editing the number of shard in a validator internal config file.

## Proposal

Create a command `linera-server edit-shards --server file.json --num-shards 4 --host 'shard-%' --port 8000 --metrics-host 'shard-%' --metrics-port 8001`

Because of port numbers, I figured we should use `%` when `num_shards` is written with 1 digits, `%%` for 2 digits, etc.

## Test Plan

```
linera-server generate --validators configuration/local/validator_*.toml --committee committee.json --testing-prng-seed 1

linera-server edit-shards --server server_1.json --num-shards 02 --host 'shard-%%' --port '10000' --metrics-host 'shard-%%' --metrics-port '110%%'
```

## Release Plan

- These changes should be backported to the latest `devnet` branch, then
    - be released in a new SDK,
- These changes should be backported to the latest `testnet` branch, then
    - be released in a new SDK,

### CODE REVIEW

File: linera-service/src/server.rs [hunk 1]
```diff
@@ -395,6 +395,36 @@ enum ServerCommand {
...
+        num_shards: u8,
+
+        /// The host of the validator (IP address or hostname). `%%` will be replace by
+        /// the shard number.
+        #[arg(long)]
+        host: String,
+
+        /// The port of the main endpoint.
+        #[arg(long)]
+        port: u16,
```
Comment (christos-h) [inline linera-service/src/server.rs:417]: I prefer [NonZeroUsize](https://doc.rust-lang.org/std/num/type.NonZeroUsize.html) for ports.

File: linera-service/src/server.rs [hunk 2]
```diff
@@ -558,6 +590,36 @@ async fn run(options: ServerOptions) {
...
+        ServerCommand::EditShards {
+            server_config_path,
+            num_shards,
+            host,
+            port,
+            metrics_host,
+            metrics_port,
+        } => {
+            assert!(
+                num_shards > 0 && num_shards < 100,
```
Comment (christos-h) [inline linera-service/src/server.rs:603]: Using `NonZeroUsize` here as well will remove the need for an explicit check.

File: linera-service/src/server.rs [hunk 3]
```diff
@@ -558,6 +590,36 @@ async fn run(options: ServerOptions) {
...
+                "The number of shards should be between 1 and 99"
+            );
+            let mut server_config: ValidatorServerConfig =
+                util::read_json(&server_config_path).expect("Fail to read server config");
+            let mut shards = Vec::new();
+            for i in 0..num_shards {
+                let index = format!("{i:2}");
+                let shard = ShardConfig {
+                    host: host.replace("%%", &index),
```
Comment (christos-h) [inline linera-service/src/server.rs:612]: Unfortunately, Docker compose begins indexing service replicas from 1. So the first shard is `docker-shard-1`.

Comment (ma2bd) [reply linera-service/src/server.rs:612]: Good catch. I meant to start at `1`

File: linera-service/src/server.rs [hunk 4]
```diff
@@ -395,6 +395,36 @@ enum ServerCommand {
...
+    /// Replaces the configurations of the shards.
+    #[command(name = "edit-shards")]
+    EditShards {
+        /// Path to the file containing the server configuration of this Linera validator.
+        #[arg(long = "server")]
+        server_config_path: PathBuf,
+
+        /// The number N of shard configs to generate (1 <= N < 100).
+        #[arg(long)]
+        num_shards: u8,
```
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-service/src/server.rs]: fixed

File: linera-service/src/server.rs (hunk 1 above)
Comment (Twey) [reply linera-service/src/server.rs]: I think [`NonZeroU16`](https://doc.rust-lang.org/std/num/type.NonZeroU16.html) is a better fit.

File: linera-service/src/server.rs [hunk 5]
```diff
@@ -461,10 +495,12 @@ async fn run(options: ServerOptions) {
             max_stream_queries,
             cache_size,
         } => {
+            linera_version::VERSION_INFO.log();
+
             let genesis_config: GenesisConfig =
-                util::read_json(&genesis_config_path).expect("Fail to read initial chain config");
```
Comment (Twey) [inline linera-service/src/server.rs:465]: Strictly, these `expect`s are incorrect anyway: the string given to `expect` should be a justification for why you think the thing _won't_ fail.

File: linera-service/src/server.rs [hunk 6]
```diff
@@ -558,6 +594,50 @@ async fn run(options: ServerOptions) {
...
+                util::read_json(&server_config_path).expect("Failed to read server config");
+            let mut shards = Vec::new();
+            let len = num_shards.len();
+            let num_shards = num_shards
+                .parse::<NonZeroU16>()
+                .expect("Failed to parse the number of shards");
+            let pattern = "%".repeat(len);
+
+            for i in 1u16..=num_shards.into() {
+                let index = format!("{i:0len$}", len = len);
```
Comment (Twey) [inline linera-service/src/server.rs:616]: `len = len` is redundant, I think.

Comment (Twey) [inline linera-service/src/server.rs:616]: `len = len` is unnecessary, I think.

File: linera-service/src/server.rs [hunk 7]
```diff
@@ -558,6 +594,50 @@ async fn run(options: ServerOptions) {
...
+                let host = host.clone().replace(&pattern, &index);
```
Comment (Twey) [inline linera-service/src/server.rs:617]: `replace` will already clone (and below).

File: linera-service/src/server.rs [hunk 8]
```diff
@@ -558,6 +594,50 @@ async fn run(options: ServerOptions) {
...
+                    host,
+                    port,
+                    metrics_host,
+                    metrics_port,
+                };
+                shards.push(shard);
+            }
+            server_config.internal_network.shards = shards;
+            util::write_json(&server_config_path, &server_config)
+                .expect("Failed to write updated server config");
```
Comment (Twey) [inline linera-service/src/server.rs:639]: I'm a bit worried about non-atomic saving on error conditions.  If you use `Persist`, an atomic rename will be applied automatically.

File: linera-service/src/server.rs (hunk 5 above)
Comment (ma2bd) [reply linera-service/src/server.rs:465]: I agree but we seem to have made an exception in CLI-related panics so far.

File: linera-service/src/server.rs (hunk 6 above)
Comment (ma2bd) [reply linera-service/src/server.rs]: indeed

File: linera-service/src/server.rs (hunk 8 above)
Comment (ma2bd) [reply linera-service/src/server.rs:612]: Happy to continue the discussion separately.

**Review by Twey — APPROVED** [review]
Looks good to me :)

File: linera-service/src/server.rs [hunk 9]
```diff
@@ -461,10 +495,12 @@ async fn run(options: ServerOptions) {
             max_stream_queries,
             cache_size,
         } => {
+            linera_version::VERSION_INFO.log();
+
```
Comment (ma2bd) [inline linera-service/src/server.rs:499]: Outch. This was a bad idea

**Pull request merged**